File: ios/Core/GateMobileBridge.swift
import Foundation

/// Shared by the framework adapters. Only JSON strings cross isolation boundaries.
@MainActor
final class GateMobileBridge {
    private var clients: [String: GateAIClient] = [:]

    func execute(_ payload: String) async -> String {
        do {
            guard let args = try JSONSerialization.jsonObject(with: Data(payload.utf8)) as? [String: Any] else {
                throw GateAIError.configuration("Expected a JSON object.")
            }
            let result = try await dispatch(args)
            return encode(["result": result])
        } catch let error as GateAIError {
            if case let .server(status, detail, headers) = error {
                var fields: [String: Any] = ["code": "http", "message": "HTTP \(status)", "status": status, "headers": headers ?? [:]]
                if let detail, let data = try? JSONEncoder().encode(detail) {
                    fields["body"] = String(decoding: data, as: UTF8.self)
                }
                return encode(["error": fields])
            }
            let code: String
            switch error {
            case .configuration: code = "configuration"
            case .network: code = "network"
            case .attestationUnavailable, .attestationFailed, .secureEnclaveUnavailable: code = "attestation"
            default: code = "native"
            }
            return encode(["error": ["code": code, "message": error.localizedDescription]])
        } catch {
            return encode(["error": ["code": "native", "message": error.localizedDescription]])
        }
    }

    private func dispatch(_ args: [String: Any]) async throws -> Any {
        let operation = try string(args, "operation")
        if operation == "configure" {
            guard let config = args["configuration"] as? [String: Any],
                  let ios = config["ios"] as? [String: Any] else {
                throw GateAIError.configuration("iOS configuration is required on iOS.")
            }
            let base = try string(config, "baseUrl")
            guard let url = URLComponents(string: base), url.scheme == "https", url.host != nil,
                  url.user == nil, url.password == nil, url.query == nil, url.fragment == nil,
                  url.path.isEmpty || url.path == "/" else {
                throw GateAIError.configuration("baseUrl must be an HTTPS origin.")
            }
            let configuration = try GateAIConfiguration(
                baseURLString: base, bundleIdentifier: ios["bundleIdentifier"] as? String,
                teamIdentifier: try string(ios, "teamIdentifier"))
            let id = UUID().uuidString
            clients[id] = GateAIClient(configuration: configuration)
            return ["clientId": id]
        }
        let id = try string(args, "clientId")
        guard let client = clients[id] else { throw GateAIError.configuration("Unknown or disposed client.") }
        switch operation {
        case "dispose":
            clients.removeValue(forKey: id)
            return NSNull()
        case "clearCachedState":
            await client.clearCachedState()
            return NSNull()
        case "request", "authorizationHeaders":
            let path = try string(args, "path")
            guard path.range(of: #"^[A-Za-z0-9_~-]+(?:/[A-Za-z0-9_~.-]+)*$"#, options: .regularExpression) != nil,
                  !path.split(separator: "/").contains(where: { $0 == "." || $0 == ".." }),
                  let method = HTTPMethod(rawValue: try string(args, "method")) else {
                throw GateAIError.configuration("Invalid relative path or HTTP method.")
            }
            let headers = args["headers"] as? [String: String] ?? [:]
            for (name, value) in headers {
                guard !["authorization", "dpop", "host", "content-length", "connection"].contains(name.lowercased()),
                      name.range(of: #"^[!#$%&'*+.^_`|~0-9A-Za-z-]+$"#, options: .regularExpression) != nil,
                      !value.contains("\r"), !value.contains("\n") else {
                    throw GateAIError.configuration("Invalid or reserved request header.")
                }
            }
            if operation == "authorizationHeaders" {
                var result = try await client.authorizationHeaders(for: path, method: method, nonce: args["nonce"] as? String)
                // Case-insensitive per-request overrides, excluding protected auth headers above.
                for (key, value) in headers {
                    for existing in Array(result.keys) where existing.lowercased() == key.lowercased() { result.removeValue(forKey: existing) }
                    result[key] = value
                }
                return result
            }
            let body = (args["body"] as? String).map { Data($0.utf8) }
            guard method != .get || body == nil else { throw GateAIError.configuration("GET cannot have a body.") }
            let (data, response) = try await client.performProxyRequest(path: path, method: method, body: body, additionalHeaders: headers)
            let responseHeaders = response.allHeaderFields.reduce(into: [String: String]()) { result, item in
                result[String(describing: item.key)] = String(describing: item.value)
            }
            return ["status": response.statusCode, "headers": responseHeaders, "body": String(decoding: data, as: UTF8.self)]
        default:
            throw GateAIError.configuration("Unknown operation.")
        }
    }

    private func string(_ args: [String: Any], _ key: String) throws -> String {
        guard let value = args[key] as? String, !value.isEmpty else { throw GateAIError.configuration("Missing \(key).") }
        return value
    }

    private func encode(_ object: [String: Any]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: object) else {
            return #"{"error":{"code":"invalid_response","message":"Could not encode native response."}}"#
        }
        return String(decoding: data, as: UTF8.self)
    }
}

File: native/android/com/gateai/mobile/GateMobileBridge.kt
package com.gateai.mobile

import android.content.Context
import com.gateai.sdk.core.GateAIClient
import com.gateai.sdk.core.GateAIConfiguration
import com.gateai.sdk.core.HttpMethod
import com.gateai.sdk.network.GateApiException
import kotlinx.coroutines.CancellationException
import org.json.JSONObject
import java.net.URI
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

internal class GateMobileBridge(private val context: Context) {
    private val clients = ConcurrentHashMap<String, GateAIClient>()

    suspend fun execute(payload: String): String = try {
        JSONObject().put("result", dispatch(JSONObject(payload))).toString()
    } catch (e: CancellationException) {
        throw e
    } catch (e: GateApiException) {
        failure("http", "HTTP ${e.statusCode}", e.statusCode, e.headers, e.body)
    } catch (e: IllegalArgumentException) {
        failure("configuration", e.message ?: "Invalid configuration.")
    } catch (e: Exception) {
        failure("native", e.message ?: "Native Gate/AI operation failed.")
    }

    private suspend fun dispatch(args: JSONObject): Any {
        val operation = args.getString("operation")
        if (operation == "configure") {
            val config = args.getJSONObject("configuration")
            val android = config.optJSONObject("android") ?: error("Android configuration is required on Android.")
            val base = config.getString("baseUrl")
            val uri = URI(base)
            require(uri.scheme == "https" && uri.host != null && uri.userInfo == null && uri.query == null && uri.fragment == null && (uri.path.isNullOrEmpty() || uri.path == "/")) { "baseUrl must be an HTTPS origin." }
            val fingerprint = android.getString("signingCertSha256")
            require(fingerprint.replace(":", "").matches(Regex("^[0-9a-fA-F]{64}$"))) { "Invalid signing certificate fingerprint." }
            val project = android.stringOrNull("cloudProjectNumber")
            require(project == null || project.matches(Regex("^[1-9][0-9]{0,17}$"))) { "Invalid cloudProjectNumber." }
            val configuration = GateAIConfiguration(
                baseUrl = base, packageName = context.packageName, signingCertSha256 = fingerprint,
                cloudProjectNumber = project?.toLong(), developmentToken = android.stringOrNull("developmentToken"),
                deviceIdentifierEnabled = android.optBoolean("deviceIdentifierEnabled", false),
                logLevel = GateAIConfiguration.LogLevel.OFF
            )
            val id = UUID.randomUUID().toString()
            clients[id] = GateAIClient.create(context.applicationContext, configuration)
            return JSONObject().put("clientId", id)
        }
        val id = args.getString("clientId")
        val client = requireNotNull(clients[id]) { "Unknown or disposed client." }
        when (operation) {
            "dispose" -> { clients.remove(id); return JSONObject.NULL }
            "clearCachedState" -> { client.clearCachedState(); return JSONObject.NULL }
            "request", "authorizationHeaders" -> {
                val path = args.getString("path")
                require(path.matches(Regex("^[A-Za-z0-9_~-]+(?:/[A-Za-z0-9_~.-]+)*$")) && path.split('/').none { it == "." || it == ".." }) { "Invalid relative path." }
                val method = HttpMethod.valueOf(args.getString("method"))
                require(method == HttpMethod.GET || method == HttpMethod.POST) { "Only GET and POST are supported." }
                val rawHeaders = args.optJSONObject("headers") ?: JSONObject()
                val headers = rawHeaders.keys().asSequence().associateWith { rawHeaders.getString(it) }
                headers.forEach { (name, value) ->
                    require(name.lowercase() !in listOf("authorization", "dpop", "host", "content-length", "connection") && name.matches(Regex("^[!#$%&'*+.^_`|~0-9A-Za-z-]+$")) && !value.contains('\r') && !value.contains('\n')) { "Invalid or reserved request header." }
                }
                if (operation == "authorizationHeaders") {
                    val result = client.authorizationHeaders(path, method, args.stringOrNull("nonce")).toMutableMap()
                    headers.forEach { (key, value) ->
                        result.keys.filter { it.equals(key, ignoreCase = true) }.forEach { result.remove(it) }
                        result[key] = value
                    }
                    return JSONObject(result as Map<*, *>)
                }
                val body = args.stringOrNull("body")?.toByteArray(Charsets.UTF_8)
                require(method != HttpMethod.GET || body == null) { "GET cannot have a body." }
                val response = client.performProxyRequest(path, method, body, headers)
                return JSONObject().put("status", response.status).put("headers", JSONObject(response.headers as Map<*, *>)).put("body", response.body ?: "")
            }
            else -> throw IllegalArgumentException("Unknown operation.")
        }
    }

    private fun failure(code: String, message: String, status: Int? = null, headers: Map<String, String> = emptyMap(), body: String? = null): String =
        JSONObject().put("error", JSONObject().put("code", code).put("message", message)
            .put("status", status ?: JSONObject.NULL).put("headers", JSONObject(headers as Map<*, *>))
            .put("body", body ?: JSONObject.NULL)).toString()

    private fun JSONObject.stringOrNull(key: String): String? = if (has(key) && !isNull(key)) getString(key) else null
}
